arctic/serialization/numpy_records.py

```diff
@@ -138,15 +148,48 @@ def _to_records(self, df, string_max_len=None, forced_dtype=None):
 
         return (rtn, dtype)
 
+    def fast_check_serializable(self, df):
+        """
+        Convert efficiently the frame's object-columns/object-index/multi-index/multi-column to
+        records, by creating a recarray only for the object fields instead for the whole dataframe.
+        If we have no object dtypes, we can safely convert only the first row to recarray to test if serializable.
+
+        Previously we'd serialize twice the full dataframe when it included object fields or multi-index/columns.
+
+        Parameters
+        ----------
+        df: `pandas.DataFrame` or `pandas.Series`
+
+        Returns
+        -------
+        `tuple[numpy.core.records.recarray, dict[str, numpy.dtype]`
+            If any object dtypes are detected in columns or index will return a dict with field-name -> dtype
+             mappings, and empty dict otherwise.
+        """
+        i_dtype, f_dtypes = df.index.dtype, df.dtypes
+
+        index_has_object = df.index.dtype.hasobject
+        fields_with_object = [f for f in df.columns if f_dtypes[f].hasobject]
+
+        if df.empty or (not index_has_object and not fields_with_object):
+            arr, _ = self._to_records(df.iloc[:1])  # only first row for performance
+            return arr, {}
+
+        # If only the Index has Objects,
+        # choose a small slice (two columns if possible, to avoid switching from a DataFrame to a Series)
+        df_objects_only = df[fields_with_object if fields_with_object else df.columns[:2]]
+
+        # Let any exceptions bubble up from here
+        arr, dtype = self._to_records(df_objects_only)
+        return arr, {f: dtype[f] for f in dtype.names}
```

Comment:
I'm a bit concerned that the arr returned here could potentially pass one of the conditional checks in can_convert_to_records_without_objects when they really shouldn't.

Reply:
Is there a simple example of such a case which I could code in the form of a unit test ?
Which of the below is your main concern?

The conditional checks are two:
(1) If the resulting arr has objects (should be safe as we serialize whole columns of the df who contain objects)
```python
            if arr.dtype.hasobject:
                log.info('Pandas dataframe %s contains Objects, saving as Blob' % symbol)
                # Will fall-back to saving using Pickle
                return False
```

(2) Multi dimensional elements, we shouldn't do
```python
            elif any([len(x[0].shape) for x in arr.dtype.fields.values()]):
```

Reply:
I was mainly concerned with (2). So you are saying that if len(x[0].shape) is larger than 0, then f_dtypes[f].hasobject will always be true? I think that makes sense. Would a np.array be considered an object in this case?

Reply:
I was under the assumption that this is the case, but is a good point.
I need to add a test to verify this, and prove compatibility.

Reply:
@aflag  I added test cases to cover this, see here for more details:
https://github.com/manahl/arctic/issues/590